### ISSUE TITLE
C++ application sample does not build on M1 Mac

### ISSUE THREAD

**Issue by msridhar** [gradle/gradle#18876]
### Expected Behavior

With Gradle 7.3 RC3 I followed the instructions here for building a C++ application sample:

https://docs.gradle.org/current/samples/sample_building_cpp_applications.html

After following the steps I expected `./gradlew build` to succeed.

### Current Behavior

`./gradlew build` fails with the following error:

```
> Task :app:compileDebugCpp FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:compileDebugCpp'.
> Error while evaluating property 'compilerVersion' of task ':app:compileDebugCpp'
   > No tool chain is available to build C++ for host operating system 'Mac OS X' architecture 'arm-v8':
       - Tool chain 'visualCpp' (Visual Studio):
           - Visual Studio is not available on this operating system.
       - Tool chain 'gcc' (GNU GCC):
           - Don't know how to build for host operating system 'Mac OS X' architecture 'arm-v8'.
       - Tool chain 'clang' (Clang):
           - Don't know how to build for host operating system 'Mac OS X' architecture 'arm-v8'.
```
### Context

I am trying to update the [WALA](https://github.com/wala/WALA) project to build on M1 Macs, and this is a blocker.

### Steps to Reproduce 

Simply running `gradle init` and following the steps on [this page](https://docs.gradle.org/current/samples/sample_building_cpp_applications.html) for building a C++ application sample is sufficient to reproduce.

### Your Environment

Build scan URL: https://gradle.com/s/2soupmqxxdmxm

I have the XCode command-line tools installed.

```
$ gcc --version
Configured with: --prefix=/Library/Developer/CommandLineTools/usr --with-gxx-include-dir=/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/4.2.1
Apple clang version 13.0.0 (clang-1300.0.29.3)
Target: arm64-apple-darwin21.1.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
$ clang --version
Apple clang version 13.0.0 (clang-1300.0.29.3)
Target: arm64-apple-darwin21.1.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```

I tried with both Zulu JDK 11 and Zulu JDK 17.

Thanks for the help!

**Comment by zhangt2333** [comment]
I encountered the same problem as you when I use JVM `Zulu 8.56.0.23-CA-macos-aarch64`.

After I switched my JVM to `Temurin JVM build 1.8.0_312-b07 macOS x64` which maybe runs under Rosetta2, everything was fine.

So is there a possible bug in the Native Platform?

**Comment by msridhar** [comment]
@zhangt2333 Yes I think this is a native platform problem.  As far as I can see it is coded to only compile to x86, even if the native toolchain on the system compiles to ARM.

Thanks for the workaround of switching to Temurin and running under Rosetta2!  I'm curious, are the final binaries you get after C++ compilation ARM binaries then, or x86?

**Comment by zhangt2333** [comment]
[transitive-dependencies example](https://github.com/gradle/native-samples/tree/master/cpp/transitive-dependencies) with `Temurin JVM build 1.8.0_312-b07 macOS x64`:
```bash
$ file ./app
./app: Mach-O 64-bit executable x86_64
```

[the example you mentioned](https://docs.gradle.org/current/samples/sample_building_cpp_applications.html) with `Zulu 8.56.0.23-CA-macos-aarch64`:
```bash
$ file ./libTest
./libTest: Mach-O 64-bit executable arm64
```

**Comment by msridhar** [comment]
That's interesting!  I had misread your previous comment; I'm surprised you got the application sample to compile with Zulu.  When I tried it just now with `Zulu 8.58.0.13-CA-macos-aarch64` and Gradle 7.3 it still fails with the same error.

**Comment by msridhar** [comment]
I did some spelunking through the code (which I don't fully understand), and for GCC / Clang I think the only built-in supported architectures are x86 32- and 64-bit:

https://github.com/gradle/gradle/blob/732148799d2e7b0efe07b2649d5b80b4b3ab7c3f/subprojects/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/internal/gcc/AbstractGccCompatibleToolChain.java#L95-L96

I feel like this code might need to be modified to be conditioned on the architecture of the machine where the build is occurring, adding ARM options.  But again, I don't really know this code.

**Comment by zhangt2333** [comment]
> That's interesting! I had misread your previous comment; I'm surprised you got the application sample to compile with Zulu. When I tried it just now with `Zulu 8.58.0.13-CA-macos-aarch64` and Gradle 7.3 it still fails with the same error.

I'm so curious why you failed in that example that I retried it. I am sorry that I  have confused two examples [sample_building_cpp_applications](https://docs.gradle.org/current/samples/sample_building_cpp_applications.html) and [sample_building_cpp_libraries](https://docs.gradle.org/current/samples/sample_building_cpp_libraries.html).

I made binaries with the [sample_building_cpp_libraries](https://docs.gradle.org/current/samples/sample_building_cpp_libraries.html), and can make an arm one or an x64 one, while I failed as you with the [sample_building_cpp_applications](https://docs.gradle.org/current/samples/sample_building_cpp_applications.html).

Then I suspect the problem may be in the plugin `cpp-library` or `cpp-application`.

Another interesting thing I found is that, if I add
```groovy
application {
    targetMachines.add(machines.macOS.x86_64)
}
```
to the `build.gradle` with the [sample_building_cpp_applications](https://docs.gradle.org/current/samples/sample_building_cpp_applications.html), although I used the `Zulu 8.56.0.23-CA-macos-aarch64` to run Gradle, I made an x64 binary:
```bash
$ file ./app    
./app: Mach-O 64-bit executable x86_64
```

**Comment by msridhar** [comment]
@zhangt2333 thanks so much for digging into this! Your trick to get the build to work on Zulu allowed me to work around this issue in our project (https://github.com/wala/WALA/pull/1052).  I think adding the macOS x86_64 target machine convinces Gradle to just run the compiler with default settings, which in the end produces an ARM binary when Gradle is run on Zulu. This is a good workaround I think until this issue can be fixed. Thanks again!

**Comment by dfsutherland** [comment]
This still fails with Gradle 8.0.1 on MacOS 13.2.1 running on an M1Pro-based laptop.  The error message is as follows:
```> Task :cast:cast:linkDebug FAILED
ld: warning: ignoring file /Library/Java/JavaVirtualMachines/temurin-11.jdk/Contents/Home/lib/server/libjvm.dylib, building for macOS-x86_64 but attempting to link with file built for macOS-arm64
Undefined symbols for architecture x86_64:
  "_JNI_CreateJavaVM", referenced from:
      launch_jvm(char*) in launch.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```
My example is the WALA test from the OP.

**Comment by ljacomet** [comment]
Sorry for the late reply!

***

Could anyone try with Gradle 8.11?
This following issue might have resolved the problem:
* #17660
